src/core/team-core.js

```diff
@@ -1,24 +1,98 @@
+}
+
+function _getField(cityId) {
+  return `city_${ !cityId ? 'all' : cityId }`;
+}
+
 function getTeams(opts) {
-  const isBanned = opts.client && !!opts.client.isBanned;
+  _isStale().then(stale => {
+    // Incur a cache refresh if necessary
+    if (stale) _updateCache();
+  });
+
+  // Do not await for update to complete as it may take some time.
+  return BPromise.all([init]).then(() =>
+    redisClient.hgetAsync(KEY_TEAMS, _getField(opts.city))
+      .then(result =>JSON.parse(result))
+  );
+}
+
+function _isStale() {
+  return redisClient.getAsync(KEY_IS_STALE).then(stale => !(stale === 'false'));
+}
+
+function _isUpdating() {
+  return redisClient
+    .getAsync(KEY_IS_UPDATING)
+    .then(isUpdating => isUpdating == 'true');
+}
+
+/**
+ * Updates the cache if no update is in progress. Otherwise does nothing.
+ *
+ * To note is that checking for on going update is best effort only. Overlapping
+ * updates are possible, but should be very uncommon and not impact the
+ * generated cache in a negative way.
+ **/
+function _updateCache() {
+  const update = redisClient.setAsync(KEY_IS_UPDATING, 'true')
+    .then(() => knex('cities').select('*')
+      // Cache per city team rankings
+      .then(cities =>
+        BPromise.map(cities, city =>
+          _getTeams(city.id).then(teams =>
+            redisClient.hmsetAsync(KEY_TEAMS, _getField(city.id), JSON.stringify(teams))
+          )
+        )
+      )
+      // Cache nation wide team rankings
+      .then(() =>
+        _getTeams().then(teams =>
+            redisClient.hmsetAsync(KEY_TEAMS, _getField(), JSON.stringify(teams))
+        )
+      )
+      .then(() => redisClient.msetAsync(KEY_IS_STALE, 'false', KEY_IS_UPDATING, 'false'))
+      .then(() => redisClient.expire(KEY_IS_STALE, CACHE_TTL))
+    );
+
+  return _isUpdating().then(isUpdating => isUpdating ? BPromise.resolve() : update);
+}
+
+function _getTeams(city) {
   const voteScoreSql = `
     (SELECT
       team_id,
       SUM(value) AS value
     FROM
       (SELECT
         wilsons(
-            COUNT(CASE votes.value WHEN 1 THEN 1 ELSE null END)::int,
-            COUNT(CASE votes.value WHEN -1 THEN -1 ELSE null END)::int
-          ) * COUNT(votes)::numeric AS value,
+          SUM(CASE votes.value WHEN 1 THEN (1 - voter_biases.bias) ELSE null END)::numeric,
+          SUM(CASE votes.value WHEN -1 THEN voter_biases.bias ELSE null END)::numeric
+        ) * COUNT(votes) AS value,
         users.team_id AS team_id
-      FROM feed_items
+      FROM (
+        SELECT feed_items.*
+        FROM feed_items
+        JOIN users ON users.id = feed_items.user_id AND NOT users.is_banned
+      ) feed_items
       JOIN users ON users.id = feed_items.user_id
       LEFT JOIN votes ON votes.feed_item_id = feed_items.id
-      WHERE NOT users.is_banned
+      JOIN voter_biases ON voter_biases.user_id = votes.user_id AND voter_biases.team_id = users.team_id
```

Comment:
This should actually be `LEFT JOIN` as it relies on the results of the LEFT JOIN votes rows. Otherwise it drops rows that LEFT JOIN returns nulls.